consensus/src/data_manager.rs

```diff
@@ -0,0 +1,234 @@
+                    }
+                    Payload::Empty => {}
+                    Payload::DirectMempool(_) => {
+                        unreachable!()
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn get_data(&self, block: &Block) -> Result<Vec<SignedTransaction>, Error> {
+        if block.payload().is_none() || block.payload().unwrap().is_empty() {
+            return Ok(Vec::new());
+        }
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            if let Payload::InQuorumStore(proofs) = block.payload().unwrap() {
+
+                // let data_status = self.digest_status.entry(block.id());
+                match self.digest_status.entry(block.id()) {
+                    dashmap::mapref::entry::Entry::Occupied(mut entry) => match entry.get_mut() {
+                        DataStatus::Cached(data) => {
+                            return Ok(data.clone());
+                        }
+                        DataStatus::Requested(receivers) => {
+                            let mut vec_ret = Vec::new();
+                            debug!("QSE: waiting for data on {} receivers", receivers.len());
+                            for rx in receivers {
+                                match rx.await {
+                                    Err(_) => {
+                                        debug!("Oneshot channel to get a batch was dropped");
+                                    }
+                                    Ok(result) => match result {
+                                        Ok(data) => {
+                                            debug!("QSE: got data, len {}", data.len());
+                                            vec_ret.push(data);
+                                        }
+                                        Err(e) => {
+                                            debug!("QS: got error from receiver {:?}", e);
+                                            let new_receivers = self
+                                                .request_data(
+                                                    proofs.clone(),
+                                                    LogicalTime::new(block.epoch(), block.round()),
+                                                )
+                                                .await;
+                                            entry.replace_entry(DataStatus::Requested(
+                                                new_receivers,
+                                            ));
+                                            return Err(e);
+                                        }
+                                    },
+                                }
+                            }
+                            let ret: Vec<SignedTransaction> =
+                                vec_ret.into_iter().flatten().collect();
+                            entry.replace_entry(DataStatus::Cached(ret.clone()));
```

Comment:
We should probably add a comment here, there were two times the data was asked for right - that's why cached? (I don't even remember why)

Reply:
Yes, inside commit and compute. I will add a comment.